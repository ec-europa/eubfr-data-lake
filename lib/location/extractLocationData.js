/**
 * Extracts the following properties for `project_locations:
 * - `country_code`
 * - `region`
 * - `town`
 *
 * @param {String} input
 * @returns {Object} results
 */
const extractLocationData = input => {
  if (!input) return {};

  const separator = new RegExp(/[,/]/i); // separate either by comma or slash.

  const results = {
    countryCodes: [],
    regions: [],
    towns: [],
    raw: input,
  };

  const locations = input
    .split(';')
    .map(p => p.trim())
    .filter(l => l);

  locations.forEach(location => {
    const infoArray = location
      .split(separator)
      // Improve the separate pieces of information for each location.
      .map(i => i.trim())
      .filter(i => i);

    // Country code is always the last item.
    const code = infoArray.pop();
    results.countryCodes.push(code);

    if (infoArray.length) {
      const [town, region] = infoArray;

      if (town) results.towns.push(town);
      if (region) results.regions.push(region);
    }
  });

  return results;
};

module.exports = extractLocationData;
