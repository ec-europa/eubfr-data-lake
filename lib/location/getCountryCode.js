// Helper to translate few specific NUTS0 codes to ISO 3166-1 Alpha-2 country codes

const getCountryCode = countryCode => {
  // Do nothing if falsy
  if (!countryCode) return countryCode;

  switch (countryCode.trim().toUpperCase()) {
    case 'UK':
      return 'GB';

    // Greece
    case 'EL':
      return 'GR';

    default:
      return countryCode;
  }
};

module.exports = getCountryCode;
