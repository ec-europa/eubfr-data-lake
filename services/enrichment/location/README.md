# Enrichment location

Enrich `project_locations`.
