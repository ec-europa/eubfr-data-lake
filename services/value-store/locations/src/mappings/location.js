module.exports = {
  location: {
    properties: {
      geolocation: {
        type: 'geo_shape',
      },
    },
  },
};
