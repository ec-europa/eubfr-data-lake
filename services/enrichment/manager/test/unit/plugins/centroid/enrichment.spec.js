/**
 * @jest-environment node
 */

import plugin from '../../../../src/plugins/centroid';
import location from './location.json';

describe('Centroid enrichment plugin: empty object', () => {
  const loc = {};
  let result = {};

  beforeAll(async () => {
    result = await plugin(loc);
  });

  test('Returns empty object, untouched, no errors', () => {
    expect(result).toMatchSnapshot();
  });
});

describe('Centroid enrichment plugin: actual input', () => {
  const loc = location;
  let result = {};

  beforeAll(async () => {
    result = await plugin(loc);
  });

  test('Returns a new object with NUTs code', () => {
    expect(result).toMatchSnapshot();
  });
});
