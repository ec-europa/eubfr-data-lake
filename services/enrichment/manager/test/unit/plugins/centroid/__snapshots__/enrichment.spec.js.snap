// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Centroid enrichment plugin: depending on the following input Empty object: returns same object, untouched, no errors 1`] = `Object {}`;

exports[`Centroid enrichment plugin: depending on the following input Object with a centroid property containing latidude and logitude: returns a new object with NUTS data 1`] = `
Object {
  "address": "",
  "centroid": Object {
    "lat": 54.6380366,
    "lon": 25.2865583,
  },
  "country_code": "LT",
  "enriched": true,
  "location": Object {
    "coordinates": Array [
      25.2865583,
      54.6380366,
    ],
    "type": "Point",
  },
  "nuts": Array [
    Object {
      "code": "LT",
      "level": 0,
      "name": "LIETUVA",
      "year": "NUTS_2013",
    },
    Object {
      "code": "LT0",
      "level": 1,
      "name": "LIETUVA",
      "year": "NUTS_2013",
    },
    Object {
      "code": "LT00",
      "level": 2,
      "name": "Lietuva",
      "year": "NUTS_2013",
    },
    Object {
      "code": "LT00A",
      "level": 3,
      "name": "Vilniaus apskritis",
      "year": "NUTS_2013",
    },
  ],
  "postal_code": "03500",
  "region": "",
  "town": "Vilnius",
}
`;

exports[`Centroid enrichment plugin: depending on the following input Object with centroid property containing latidude and logitude, but no country_code, take country_code from NUTS 1`] = `
Object {
  "address": "",
  "centroid": Object {
    "lat": 54.6380366,
    "lon": 25.2865583,
  },
  "country_code": "LT",
  "enriched": true,
  "location": Object {
    "coordinates": Array [
      25.2865583,
      54.6380366,
    ],
    "type": "Point",
  },
  "nuts": Array [
    Object {
      "code": "LT",
      "level": 0,
      "name": "LIETUVA",
      "year": "NUTS_2013",
    },
    Object {
      "code": "LT0",
      "level": 1,
      "name": "LIETUVA",
      "year": "NUTS_2013",
    },
    Object {
      "code": "LT00",
      "level": 2,
      "name": "Lietuva",
      "year": "NUTS_2013",
    },
    Object {
      "code": "LT00A",
      "level": 3,
      "name": "Vilniaus apskritis",
      "year": "NUTS_2013",
    },
  ],
  "postal_code": "03500",
  "region": "",
  "town": "Vilnius",
}
`;
